overlord/install/install.go

```diff
@@ -193,3 +206,232 @@ func CheckEncryptionSupport(model *asserts.Model, tpmMode secboot.TPMProvisionMo
+	// validity check
+	if len(keyForRole) == 0 || keyForRole[gadget.SystemData] == nil || keyForRole[gadget.SystemSave] == nil {
+		return fmt.Errorf("internal error: system encryption keys are unset")
+	}
+	dataEncryptionKey := keyForRole[gadget.SystemData]
+	saveEncryptionKey := keyForRole[gadget.SystemSave]
+
+	// make note of the encryption keys
+	trustedInstallObserver.ChosenEncryptionKeys(dataEncryptionKey, saveEncryptionKey)
+
+	// XXX is the asset cache problematic from initramfs?
+	// keep track of recovery assets
+	if err := trustedInstallObserver.ObserveExistingTrustedRecoveryAssets(boot.InitramfsUbuntuSeedDir); err != nil {
+		return fmt.Errorf("cannot observe existing trusted recovery assets: %v", err)
+	}
+	if err := saveKeys(model, keyForRole); err != nil {
+		return err
+	}
+	// write markers containing a secret to pair data and save
+	if err := writeMarkers(model); err != nil {
+		return err
+	}
+	return nil
+}
+
+// writeMarkers writes markers containing the same secret to pair data and save.
+func writeMarkers(model *asserts.Model) error {
+	// ensure directory for markers exists
+	if err := os.MkdirAll(boot.InstallHostFDEDataDir(model), 0755); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(boot.InstallHostFDESaveDir, 0755); err != nil {
+		return err
+	}
+
+	// generate a secret random marker
+	markerSecret, err := randutil.CryptoTokenBytes(32)
+	if err != nil {
+		return fmt.Errorf("cannot create ubuntu-data/save marker secret: %v", err)
+	}
+
+	return device.WriteEncryptionMarkers(boot.InstallHostFDEDataDir(model), boot.InstallHostFDESaveDir, markerSecret)
+}
+
+func saveKeys(model *asserts.Model, keyForRole map[string]keys.EncryptionKey) error {
+	saveEncryptionKey := keyForRole[gadget.SystemSave]
+	if saveEncryptionKey == nil {
+		// no system-save support
+		return nil
+	}
+	// ensure directory for keys exists
+	if err := os.MkdirAll(boot.InstallHostFDEDataDir(model), 0755); err != nil {
+		return err
+	}
+	if err := saveEncryptionKey.Save(device.SaveKeyUnder(boot.InstallHostFDEDataDir(model))); err != nil {
+		return fmt.Errorf("cannot store system save key: %v", err)
+	}
+	return nil
+}
+
+// PrepareRunSystemData prepares the run system:
+// * it writes the model to ubuntu-boot
+// * sets up/copies any allowed and relevant cloud init configuration
+// * plus other details
+func PrepareRunSystemData(model *asserts.Model, gadgetDir string, perfTimings timings.Measurer) error {
+	// keep track of the model we installed
+	err := os.MkdirAll(filepath.Join(boot.InitramfsUbuntuBootDir, "device"), 0755)
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
+	}
+	err = writeModel(model, filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"))
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
+	}
+
+	// XXX does this make sense from initramfs?
```

Comment:
I think we can drop it. We have `/usr/lib/clock-epoch` in both core and core-initrd. So the clock should set itself to the latest build date of both the core snap and the initrd.

Otherwise, here it is copies the `clock` file of either core-base or core-initrd depending on whether we do install in initrd or in core snap. Which seems wrong.

Reply:
that's fine, we can drop it in a later PR though, as this is trying to mostly move preexisting code